src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,604 @@
+            return publicKey.ToKvmAddress();
+        }
+
+        private static void QuickFail(ContractEntry entry, ExecutionEnvironment env, ITxTracer txTracer)
+        {
+            // here we need to propagate back to Delta
+            env.CurrentBlock.GasUsed += (long) entry.GasLimit;
+            if (txTracer.IsTracingReceipt)
+            {
+                txTracer.MarkAsFailed(env.ExecutingAccount, (long) entry.GasLimit, Bytes.Empty, "invalid");
+            }
+        }
+
+        private void Execute(Delta delta, ITxTracer txTracer, bool readOnly)
+        {
+            StateUpdate stateUpdate = ToStateUpdate(delta);
+            foreach (PublicEntry publicEntry in delta.PublicEntries)
+            {
+                ContractEntry contractEntry = new ContractEntry();
+                contractEntry.Base = publicEntry.Base;
+                contractEntry.Amount = publicEntry.Amount;
+                contractEntry.Data = ByteString.Empty;
+                contractEntry.GasLimit = 21000;
+                Execute(contractEntry, stateUpdate, txTracer, readOnly);
+            }
+
+            // revert state if any fails (take snapshot)
+            foreach (ContractEntry contractEntry in delta.ContractEntries)
+            {
+                Execute(contractEntry, stateUpdate, txTracer, readOnly);
+            }
+
+            if (!readOnly)
+            {
+                // we should assign block rewards here (or in Ledger)
+                _stateProvider.CommitTree();
+                _storageProvider.CommitTrees();
+            }
+            else
+            {
+                _storageProvider.Reset();
+                _stateProvider.Reset();
+            }
+        }
+
+        private void TraceLogInvalidTx(ContractEntry entry, string reason)
+        {
+            if (_logger.IsEnabled(LogEventLevel.Verbose))
+            {
+                _logger.Verbose("Invalid enrty {entry} ({reason})", entry, reason);
+            }
+        }
+
+        /// <summary>
+        /// Executes the <paramref name="entry"/>.
+        /// </summary>
+        /// <param name="entry">Transaction entry to be executed inside the <see cref="VirtualMachine"/>.</param>
+        /// <param name="stateUpdate"><see cref="Delta"/> to be used for execution environment construction</param>
+        /// <param name="txTracer">Tracer to extract the execution steps for debugging or analytics.</param>
+        /// <param name="readOnly">Defines whether the state should be reverted after the execution.</param>
+        /// <exception cref="TransactionCollisionException">Thrown when deployment address already has some code.</exception>
+        /// <exception cref="OutOfGasException">Thrown when not enough gas is available for deposit.</exception>
+        private void Execute(ContractEntry entry, StateUpdate stateUpdate, ITxTracer txTracer, bool readOnly)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The Cyclomatic Complexity of this method is 35 which is greater than 10 authorized.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4350015)